server/webapp/WEB-INF/rails/webpack/models/new_pipeline_configs/materials.ts

```diff
@@ -36,41 +35,23 @@ import {
   TfsMaterialAttributesJSON,
 } from "./materials_serialization";
 
-//tslint:disable-next-line
-export interface Material extends ValidatableMixin {
-}
-
-//tslint:disable-next-line
-export interface MaterialAttributes extends ValidatableMixin {
-}
-
-//tslint:disable-next-line
-export interface GitMaterialAttributes extends ValidatableMixin {
-}
-
-//tslint:disable-next-line
-export interface SvnMaterialAttributes extends ValidatableMixin {
-}
-
-//tslint:disable-next-line
-export interface HgMaterialAttributes extends ValidatableMixin {
-}
-
-//tslint:disable-next-line
-export interface P4MaterialAttributes extends ValidatableMixin {
-}
-
-//tslint:disable-next-line
-export interface TfsMaterialAttributes extends ValidatableMixin {
-}
+export class Materials extends Array<Material> {
+  constructor(...materials: Material[]) {
+    super(...materials);
+    Object.setPrototypeOf(this, Object.create(Materials.prototype));
+  }
 
-//tslint:disable-next-line
-export interface DependencyMaterialAttributes extends ValidatableMixin {
-}
+  remove(material: Material) {
+    const index = this.findIndex((record) => record === material);
+    this.splice(index, 1);
+  }
 
-export class Materials extends NameableSet<Material> {
-  constructor(...materials: Material[]) {
-    super(materials);
+  isValid(): boolean {
```

Comment:
While this works, it isn't the pattern we're following in other classes. Take a look at the usages of `applyMixins`. We have a dummy interface with the same name and then the actual class implements ValidatableMixin.

something like -

```
//tslint:disable-next-line
export interface Materials extends ValidatableMixin {
}

export class Materials extends Array<Material> implements ValidatableMixin {
}
```
In this case, we can also remove `Object.setPrototypeOf(this, Object.create(Materials.prototype));` from the constructor.

I think that extending `Array<Material>` isn't giving us much. We can do the validation in the PipelineConfig class itself - we have methods available in ValidatableMixin for associated collections. Let me know what you think.

Reply:
I see what you are saying. Tried your suggestion and it worked! so kept `validateAssociated` for materials validation as suggested in [this comment](https://github.com/gocd/gocd/pull/6884#discussion_r321597469)

However, `Object.setPrototypeOf(this, Object.create(Materials.prototype));` is required because otherwise the prototype of `Materials` class remains an Array.

Reply:
hmm...that's interesting 🤔